src/transformStateWithClones.js

```diff
@@ -7,7 +7,35 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  for (const { type, keysToRemove, extraData } of actions) {
+    switch (type) {
+      case 'addProperties':
+        addProps(currentState, extraData);
+        break;
+      case 'removeProperties':
+        removeProps(currentState, keysToRemove);
+        break;
+      default:
+        currentState = {};
+    }
+
+    stateHistory.push({ ...currentState });
+  }
+
+  function addProps(obj, extraData) {
+    Object.assign(obj, extraData);
```

Comment:
The function `addProps` should not mutate the passed object. Instead, it should create a new copy with the added properties and return it. This can be done using the spread operator or `Object.assign` with a new object as the first parameter.